### ISSUE TITLE
Progress report

### ISSUE THREAD

**Issue by kan-bayashi** [kan-bayashi/ParallelWaveGAN#1]
This issue records the progress of training.

**Comment by kan-bayashi** [comment]
60k iters

<img width="896" alt="スクリーンショット 2019-10-31 17 57 22" src="https://user-images.githubusercontent.com/22779813/67933003-4f301580-fc08-11e9-93cd-a12e32068f82.png">
<img width="946" alt="スクリーンショット 2019-10-31 17 57 50" src="https://user-images.githubusercontent.com/22779813/67933016-5525f680-fc08-11e9-84cc-dfefb07c2b63.png">

**Comment by kan-bayashi** [comment]
I finished the almost all of the implementation.
I will start the training based on the recipe.
https://github.com/kan-bayashi/ParallelWaveGAN/blob/master/egs/ljspeech/voc1/run.sh

**Comment by kan-bayashi** [comment]
75k samples are available.
https://drive.google.com/open?id=1sd_QzcUNnbiaWq7L0ykMP7Xmk-zOuxTi

The training curve is as follows:
<img width="1150" alt="スクリーンショット 2019-11-02 午前11 32 26" src="https://user-images.githubusercontent.com/22779813/68064887-1063a200-fd65-11e9-9312-b6cc07b85e20.png">

Interestingly, GAN is not applied yet but the quality is not so bad.
WaveNet architecture and STFT-based loss are very strong.

**Comment by rishikksh20** [comment]
@kan-bayashi looks promising. I have one vacant v100, I definitely going to start training this weekend.

**Comment by kan-bayashi** [comment]
Hi @rishikksh20.
That is nice. 
If you find any important points, please share with me.

**Comment by kan-bayashi** [comment]
Added 100k and 130k samples.
https://drive.google.com/open?id=1sd_QzcUNnbiaWq7L0ykMP7Xmk-zOuxTi
It seems the quality is improving gradually.

**Comment by rishikksh20** [comment]
@kan-bayashi Yes sound quality improved lot, without headphones its hard to judge. I am also started fresh training to determine several parameters for TTS baseline. 
What is inference speed, is it real time ?

**Comment by kan-bayashi** [comment]
@rishikksh20 The speed is amazing. I think this is enough.
```bash
2019-11-03 09:07:03,286 (decode:89) INFO: the number of features to be decoded = 250.
2019-11-03 09:07:10,390 (decode:103) INFO: loaded model parameters from exp/train_ljspeech_parallel_wavegan.v1/checkpoint-175000steps.pkl.
[decode]: 100%|██████████| 250/250 [00:30<00:00,  8.31it/s, RTF=0.0156]
2019-11-03 09:07:40,480 (decode:127) INFO: finished generation of 250 utterances (RTF = 0.016).
```

**Comment by kan-bayashi** [comment]
Added 200k samples.
https://drive.google.com/open?id=1sd_QzcUNnbiaWq7L0ykMP7Xmk-zOuxTi
I think it reached a nice quality.

I attached the training curve.
Pulse in 120k is caused by resuming. Please ignore it.

### Spectral convergence loss
<img width="1260" alt="スクリーンショット 2019-11-03 午後3 29 46" src="https://user-images.githubusercontent.com/22779813/68081311-644ab580-fe4f-11e9-87d0-2f00467d1fb5.png">

### log STFT magnitude loss
<img width="1262" alt="スクリーンショット 2019-11-03 午後3 29 57" src="https://user-images.githubusercontent.com/22779813/68081313-70367780-fe4f-11e9-9d2c-8e6b0dabedee.png">

### adversarial loss
<img width="1274" alt="スクリーンショット 2019-11-03 午後3 31 05" src="https://user-images.githubusercontent.com/22779813/68081321-7debfd00-fe4f-11e9-96db-6a5e23d24d4b.png">

### Generator loss (sc + mag + adv)
<img width="1261" alt="スクリーンショット 2019-11-03 午後3 31 40" src="https://user-images.githubusercontent.com/22779813/68081329-93f9bd80-fe4f-11e9-8757-cf4ecc156fe5.png">

### fake loss
<img width="1271" alt="スクリーンショット 2019-11-03 午後3 31 58" src="https://user-images.githubusercontent.com/22779813/68081333-a96ee780-fe4f-11e9-921f-a9740dcda58e.png">

### real loss
<img width="1268" alt="スクリーンショット 2019-11-03 午後3 32 23" src="https://user-images.githubusercontent.com/22779813/68081335-b25fb900-fe4f-11e9-8d5b-a676a9163ff1.png">

### Discriminator loss (fake + real)
<img width="1270" alt="スクリーンショット 2019-11-03 午後3 31 17" src="https://user-images.githubusercontent.com/22779813/68081339-c0153e80-fe4f-11e9-8b07-e8dc157c31aa.png">

**Comment by kan-bayashi** [comment]
I made Mandarin and Japanese recipes.
I will start to train them.

**Comment by kan-bayashi** [comment]
Added 280k samples.
Now it reached almost the same quality as the official samples.
https://r9y9.github.io/demos/projects/icassp2020/

**Comment by kan-bayashi** [comment]
Added jsut and csmsc intermediate samples.

**Comment by kan-bayashi** [comment]
I added arctic recipe. #24
If it is possible to train with only 1000 utterances, that is great news for the voice conversion.
I will try to train.

**Comment by xzm2004260** [comment]
@kan-bayashi  The original model is 1.4MB, why is the model size 16.7MB now?

**Comment by kan-bayashi** [comment]
@xzm2004260 This is because the checkpoint includes all of the states of the models, optimizers, and schedulers.

**Comment by kan-bayashi** [comment]
I am comparing two models, one follows the paper suggestion to start to train the discriminator from 100k iters, and the other starts to train from the first step.
<img width="1627" alt="スクリーンショット 2019-11-04 午後7 47 52" src="https://user-images.githubusercontent.com/22779813/68115567-0647cc00-ff3c-11e9-8c1a-c20189dd43d4.png">
From the training curve, the discriminator can be trained from the first step.
And in terms of log Magnitude STFT loss, the model trained from the first step is better while spectral convergence loss is almost the same.

**Comment by kobenaxie** [comment]
@kan-bayashi How to train the model with the specified GPU ? and will support Multi GPUs training ?

**Comment by kan-bayashi** [comment]
@kobenaxie You can specify the GPU via `CUDA_VISIBLE_DEVICES` e.g.

```bash
CUDA_VISIBLE_DEVICES=1 ./run.sh --stage 2
```

It can be supported but currently, I do not feel necessity because it can train with 12 GB single GPU and it finishes within ~3 days.

**Comment by kobenaxie** [comment]
@kan-bayashi thank you for your reply very very much ~

**Comment by kan-bayashi** [comment]
I added how-to-specify the gpu in README.
https://github.com/kan-bayashi/ParallelWaveGAN#run

**Comment by kan-bayashi** [comment]
I made demo HP.
https://kan-bayashi.github.io/ParallelWaveGAN/

**Comment by kan-bayashi** [comment]
I added intermediate results of jsut, csmsc, and arctic.

**Comment by MorganCZY** [comment]
I run this repo with LJSpeech on a 2080Ti gpu. When the code runs at 1000 steps, it raises an error.
![image](https://user-images.githubusercontent.com/33675916/68187626-187c4580-ffe2-11e9-85dd-abd89d3cb7c6.png)
Could you give me some advice on how to fix this?

**Comment by kan-bayashi** [comment]
@MorganCZY 
The easiest way is to use a little bit smaller batch_max_steps.
(Make sure it should be the multiple of hop_size.

Could you give me the exact gpu memory size of 2080 Ti? 
In my case, it requires 11613 MiB memory, so I think 12 GB memory is enough.

**Comment by kan-bayashi** [comment]
I found a bug in the evaluation.
I forgot to add `torch.no_grad()` in evaluation, causing additonal memory comsuming.
I will fix it.

**Comment by MorganCZY** [comment]
![image](https://user-images.githubusercontent.com/33675916/68188956-52028000-ffe5-11e9-9ea3-3150a0dd83e6.png)
I used gpu 0 shown in above picture. In training stage, the memory is enough. But after 1000 steps, it steps in eval process and the error occurs.

**Comment by kan-bayashi** [comment]
Try #26.

**Comment by MorganCZY** [comment]
> I found a bug in the evaluation.
> I forgot to add `torch.no_grad()` in evaluation, causing additonal memory comsuming.
> I will fix it.

That makes sense for my errors

**Comment by kan-bayashi** [comment]
After fixing (#26), the gpu memory usage is 11613 -> 11151 MiB.

**Comment by MorganCZY** [comment]
@kan-bayashi The error fixes using #26 ! Thx~

**Comment by kan-bayashi** [comment]
I summarized the comparison https://github.com/kan-bayashi/ParallelWaveGAN/issues/1#issuecomment-549301785 in #27, and found the issue of the discriminator.

**Comment by kan-bayashi** [comment]
I uploaded the initial model of csmsc and jsut.
https://drive.google.com/open?id=1sd_QzcUNnbiaWq7L0ykMP7Xmk-zOuxTi

**Comment by kan-bayashi** [comment]
I updated demo HP.

**Comment by kan-bayashi** [comment]
I finished arctic training. 
Surprisingly, Parallel WaveGAN can be trained with only 988 utterances.
drive.google.com/open?id=1sd_QzcUNnbiaWq7L0ykMP7Xmk-zOuxTi

**Comment by kan-bayashi** [comment]
Thanks to @erogol, now single node multi-process distributed multi-gpu training is supported (#30, #31).
If you want to try, please follow the instruction in README.
https://github.com/kan-bayashi/ParallelWaveGAN#run

**Comment by kan-bayashi** [comment]
Now I think single speaker model works well for various dataset. I will consider to implement multi speaker model.

**Comment by kan-bayashi** [comment]
I made the realtime E2E-TTS demonstration notebook.
You can try online with Google Colab.
English and Japanese models are available!
https://colab.research.google.com/gist/kan-bayashi/bc65baf72faaebe5efd601b013e07342/e2e-tts-demo.ipynb
I will add a Mandarin example tomorrow.

**Comment by rishikksh20** [comment]
@kan-bayashi Is it possible to train it on GTA from espnet's tacotron or fastspeech ?

**Comment by kan-bayashi** [comment]
@rishikksh20 It can be. Just replacing normalized feats with generated ones is OK.

**Comment by kan-bayashi** [comment]
I updated notebook to select tacotron2, transformer, and fastspeech!
https://colab.research.google.com/github/espnet/notebook/blob/master/tts_realtime_demo.ipynb

**Comment by kan-bayashi** [comment]
I added colab link. You can enjoy the realtime synthesis!

**Comment by kan-bayashi** [comment]
I added the Mandarin TTS example.
But I cannot understand Mandarin.
Maybe the text frontend part has some bugs.

**Comment by kan-bayashi** [comment]
This implementation became stable.
I will close this issue.

**Comment by vjdtao** [comment]
I cloned the latest git master version, and run ljspeech with parallel_wavegan.v1.yaml on a single 2080ti (12G mem), and got the exact same issue as @MorganCZY. I checked the fix #26, but the code seems reverted in the current version of parallel_wavegan/bin/train.py. Any solutions to fix?

**Comment by kan-bayashi** [comment]
@vjdtao This is different. From 100001 iters, discriminator is introduced in the trainings. Therefore the requires memory size will be increased.

**Comment by vjdtao** [comment]
@kan-bayashi thanks for the reply, I'll change the batch_size in the config file.
Should I start the training from scratch or resume from 100000steps?
2080Ti has 12G mem, how much mem does Titan V have?

**Comment by kan-bayashi** [comment]
Titan V has 12036 MB memory.
You can resume from 100k steps if you do not care about the reproducibility.

**Comment by vjdtao** [comment]
@kan-bayashi After tuned the batch_size down a bit, the training was finished successfully. Thanks much!
Also sorry for the incorrect memory size of 2080Ti in my previous comment, the correct size should be 11G.

**Comment by chikiuso** [comment]
Hi @kan-bayashi , sorry to reopen this issue again , I see you make some code change after people mentioning the lack of memory of rtf 2080 ti (11G), is it now solved? could I train 400k steps with 2080 ti card now? thanks.

**Comment by kan-bayashi** [comment]
I did not check since I do not hvae 2080ti, but required memory is not changed.
Please make the batch small if OOM happens.

**Comment by chikiuso** [comment]
thanks for your advice!
